File: library/src/main/java/com/rjsvieira/recyclerviewmultipleviewtypesadapter/EnumMapBindAdapter.java
package com.rjsvieira.recyclerviewmultipleviewtypesadapter;

import java.util.HashMap;
import java.util.Map;

/**
 * Adapter class for managing data binders by mapping enum view type and data binder
 * <p>
 * Created by yqritc on 2015/03/25, modified by rjsvieira.
 */

public abstract class EnumMapBindAdapter<E extends Enum<E>> extends DataBindAdapter {

    private Map<E, DataBinder> binderMap = new HashMap<>();

    @Override
    public int getItemCount() {
        int itemCount = 0;
        if (this.binderMap != null && !this.binderMap.isEmpty()) {
            for (DataBinder binder : binderMap.values()) {
                itemCount += binder.getItemCount();
            }
        }
        return itemCount;
    }

    @Override
    public int getItemViewType(int position) {
        return getEnumFromPosition(position).ordinal();
    }

    @Override
    public <T extends DataBinder> T getDataBinder(int viewType) {
        return getDataBinder(getEnumFromOrdinal(viewType));
    }

    @Override
    public int getPosition(DataBinder binder, int binderPosition) {
        E targetViewType = getEnumFromBinder(binder);
        for (int i = 0, count = getItemCount(); i < count; i++) {
            if (targetViewType != null && targetViewType == getEnumFromPosition(i)) {
                binderPosition--;
                if (binderPosition < 0) {
                    return i;
                }
            }
        }
        return getItemCount();
    }

    @Override
    public int getBinderPosition(int position) {
        E targetViewType = getEnumFromPosition(position);
        int binderPosition = -1;
        for (int i = 0; i <= position; i++) {
            if (targetViewType == getEnumFromPosition(i)) {
                binderPosition++;
            }
        }
        return binderPosition;
    }

    @Override
    public void notifyBinderItemRangeChanged(DataBinder binder, int positionStart, int itemCount) {
        for (int i = positionStart; i <= itemCount; i++) {
            notifyItemChanged(getPosition(binder, i));
        }
    }

    @Override
    public void notifyBinderItemRangeInserted(DataBinder binder, int positionStart, int itemCount) {
        for (int i = positionStart; i <= itemCount; i++) {
            notifyItemInserted(getPosition(binder, i));
        }
    }

    @Override
    public void notifyBinderItemRangeRemoved(DataBinder binder, int positionStart, int itemCount) {
        for (int i = positionStart; i <= itemCount; i++) {
            notifyItemRemoved(getPosition(binder, i));
        }
    }

    @Override
    public EnumMapBindAdapter clearDataBinderAdapter() {
        if (this.binderMap != null) {
            this.binderMap.clear();
        }
        return this;
    }

    public abstract E getEnumFromPosition(int position);

    public abstract E getEnumFromOrdinal(int ordinal);

    public Map<E, DataBinder> getBinderMap() {
        return binderMap;
    }

    protected void putBinder(E enumerator, DataBinder binder) {
        if (this.binderMap != null && enumerator != null && binder != null) {
            binderMap.put(enumerator, binder);
        }
    }

    protected <T extends DataBinder> T getDataBinder(E e) {
        return this.binderMap != null ? (T) binderMap.get(e) : null;
    }

    private E getEnumFromBinder(DataBinder binder) {
        if (this.binderMap != null && !this.binderMap.isEmpty()) {
            for (Map.Entry<E, DataBinder> entry : binderMap.entrySet()) {
                if (entry.getValue().equals(binder)) {
                    return entry.getKey();
                }
            }
        }
        return null;
    }

}

File: library/src/main/java/com/rjsvieira/recyclerviewmultipleviewtypesadapter/EnumMapBindAdapter.kt
package com.rjsvieira.recyclerviewmultipleviewtypesadapter

import com.rjsvieira.recyclerviewmultipleviewtypesadapter.views.BaseViewHolder
import java.util.*

/**
 * Adapter class for managing data binders by mapping enum view type and data binder
 *
 *
 * Created by yqritc on 2015/03/25, modified by rjsvieira.
 */

abstract class EnumMapBindAdapter<E : Enum<E>> : DataBindAdapter() {

    private val binderMap : HashMap<E, DataBinder<BaseViewHolder>> = HashMap()

    override fun getItemCount(): Int {
        var itemCount = 0
        if (!this.binderMap.isEmpty()) {
            for (binder in binderMap.values) {
                itemCount += binder.itemCount
            }
        }
        return itemCount
    }

    override fun getItemViewType(position: Int): Int {
        return getEnumFromPosition(position).ordinal
    }

    override fun getDataBinder(viewType: Int): DataBinder<BaseViewHolder> {
        return getDataBinder(getEnumFromOrdinal(viewType))
    }

    public override fun getPosition(binder: DataBinder<*>, binderPosition : Int): Int {
        var newBinderPosition = binderPosition
        val targetViewType = getEnumFromBinder(binder)
        var i = 0
        val count = itemCount
        while (i < count) {
            if (targetViewType != null && targetViewType == getEnumFromPosition(i)) {
                newBinderPosition--
                if (newBinderPosition < 0) {
                    return i
                }
            }
            i++
        }
        return itemCount
    }

    public override fun getBinderPosition(position: Int): Int {
        val targetViewType = getEnumFromPosition(position)
        var binderPosition = -1
        for (i in 0..position) {
            if (targetViewType === getEnumFromPosition(i)) {
                binderPosition++
            }
        }
        return binderPosition
    }

    override fun notifyBinderItemRangeChanged(binder: DataBinder<*>, positionStart: Int, itemCount: Int) {
        for (i in positionStart..itemCount) {
            notifyItemChanged(getPosition(binder, i))
        }
    }

    override fun notifyBinderItemRangeInserted(binder: DataBinder<*>, positionStart: Int, itemCount: Int) {
        for (i in positionStart..itemCount) {
            notifyItemInserted(getPosition(binder, i))
        }
    }

    override fun notifyBinderItemRangeRemoved(binder: DataBinder<*>, positionStart: Int, itemCount: Int) {
        for (i in positionStart..itemCount) {
            notifyItemRemoved(getPosition(binder, i))
        }
    }

    override fun clearDataBinderAdapter(): EnumMapBindAdapter<*> {
            this.binderMap.clear()
        return this
    }

    abstract fun getEnumFromPosition(position: Int): E

    abstract fun getEnumFromOrdinal(ordinal: Int): E

    fun getBinderMap(): Map<E, DataBinder<BaseViewHolder>> {
        return binderMap
    }

    protected fun putBinder(enumerator: E?, binder: DataBinder<BaseViewHolder>) {
        if (enumerator != null) {
            binderMap.put(enumerator, binder)
        }
    }

    protected fun getDataBinder(e: E): DataBinder<BaseViewHolder> {
        return binderMap[e]!!
    }

    private fun getEnumFromBinder(binder: DataBinder<*>): E? {
        if (!this.binderMap.isEmpty()) {
            for ((key, value) in binderMap) {
                if (value == binder) {
                    return key
                }
            }
        }
        return null
    }

}
